numpy/lib/arraypad.py

```diff
@@ -74,648 +70,179 @@ def _round_ifneeded(arr, dtype):
         arr.round(out=arr)
 
 
-def _slice_at_axis(shape, sl, axis):
+def _slice_at_axis(shape, axis, sl):
     """
     Construct a slice tuple the length of shape, with sl at the specified axis
     """
     slice_tup = (slice(None),)
     return slice_tup * axis + (sl,) + slice_tup * (len(shape) - axis - 1)
 
 
-def _slice_first(shape, n, axis):
-    """ Construct a slice tuple to take the first n elements along axis """
-    return _slice_at_axis(shape, slice(0, n), axis=axis)
-
-
-def _slice_last(shape, n, axis):
-    """ Construct a slice tuple to take the last n elements along axis """
-    dim = shape[axis]  # doing this explicitly makes n=0 work
-    return _slice_at_axis(shape, slice(dim - n, dim), axis=axis)
-
-
-def _do_prepend(arr, pad_chunk, axis):
-    return np.concatenate(
-        (pad_chunk.astype(arr.dtype, copy=False), arr), axis=axis)
-
-
-def _do_append(arr, pad_chunk, axis):
-    return np.concatenate(
-        (arr, pad_chunk.astype(arr.dtype, copy=False)), axis=axis)
-
-
-def _prepend_const(arr, pad_amt, val, axis=-1):
-    """
-    Prepend constant `val` along `axis` of `arr`.
+def _pad_empty(arr, pad_widths):
+    """Pad array with undefined values.
 
     Parameters
     ----------
     arr : ndarray
-        Input array of arbitrary shape.
-    pad_amt : int
-        Amount of padding to prepend.
-    val : scalar
-        Constant value to use. For best results should be of type `arr.dtype`;
-        if not `arr.dtype` will be cast to `arr.dtype`.
-    axis : int
-        Axis along which to pad `arr`.
+        Array to grow.
+    pad_widths : sequence of tuple[int, int]
+        Pad width on both sides for each dimension in `arr`.
 
     Returns
     -------
-    padarr : ndarray
-        Output array, with `pad_amt` constant `val` prepended along `axis`.
-
+    padded : ndarray
+        Larger array with undefined values in padded areas.
+    old_area : tuple
+        A tuple of slices pointing to the area of the original array.
     """
-    if pad_amt == 0:
-        return arr
-    padshape = tuple(x if i != axis else pad_amt
-                     for (i, x) in enumerate(arr.shape))
-    return _do_prepend(arr, np.full(padshape, val, dtype=arr.dtype), axis)
+    # Allocate grown array
+    new_shape = tuple(
+        size + left + right
```

Comment:
nit: `left + size + right` would be clearer, since that reflects the order of the pieces too